### PR TITLE
Update dashboard to design v122 (Part 1)

### LINKED ISSUE #9886: Update the Dashboard to Newest Design
- Remove the home tab icon
- Add a new `Start with a template` button to dashboard.
- Clicking this button should open the templates (old home tab) as a pop up over the asset list.
- Add a Drag and drop area to bottom to the asset list.
- **We will move to tabs at the top as a follow up ticket.**

image

image

### PR DESCRIPTION
**Pull request by somebody1234**
### Pull Request Description
- Close #9886
  - Update style of buttons in Drive Bar
  - Move "Home" page into a "Start" dialog
  - Remove icons that are no longer needed
  - Remove Backend Switcher in top bar - they have now been converted to categories

- Incidental changes
  - Refactor Backend provider so that both Remote and Local backends are available.
    - This was done because both Cloud and Local backends are now easily accessible by switching tabs - the Local backend no longer has its own views with the hidden category switcher

### Important Notes
None

### Code review tips
- There are a lot of changes in Settings components, for example (and in general, any component that needs to use a backend). However, most of the changes are identical, so only the approach needs to be reviewed - although the settings pages and the other pages should probably be reviewed separately.

### Testing instructions
- Make sure new/changed features work, as mentioned above (drive bar, start page, drag-n-drop)
- Because of the backend changes, all flows in general should be checked, at least at a basic level:
  - Sign up/sign in
  - Offline mode (bottom link on login screen on desktop app)
  - Checking that all settings pages work, both when logged in and not logged in

### Screencasts

https://github.com/enso-org/enso/assets/4046547/d8f385e2-3355-404f-a346-035917d40cf9

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] ~~The documentation has been updated, if necessary.~~
- [x] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      [TypeScript](https://github.com/enso-org/enso/blob/develop/docs/style-guide/typescript.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- [x] ~~Unit tests have been written where possible.~~

### CODE REVIEW

**Comment by somebody1234** [comment]
I guess I'll avoid stacking the move to tabs on top of this PR, as this is a big PR so there may be a lot of changes needed.

**Comment by PabloBuchu** [comment]
1. This loader is barely moving when running cloud project

Screenshot 2024-05-20 at 16 16 55

2. After cmd+r when having project open in progress I no longer see calls `projects/<project_id>` so I had to reload again to gain access to opened project

**Comment by somebody1234** [comment]
> This loader is barely moving when running cloud project

weird, i seem to be able to reproduce on develop. seems like it might be a regression

**Comment by somebody1234** [comment]
> After cmd+r when having project open in progress I no longer see calls `projects/<project_id>`

ah. i guess we should store the "currently open project" *immediately* when the user clicks play, rather than after it actually opens. it's late though so probably won't be fixed today (and possibly will fix on the tabs PR instead.)

**Comment by MrFlashAccount** [comment]
@somebody1234 could you please rebase onto the latest changes so I can review the PR?

**Comment by somebody1234** [comment]
@MrFlashAccount latest changes pulled.

File: app/ide-desktop/lib/dashboard/src/components/AriaComponents/Dialog/Dialog.tsx [hunk 1]
```diff
@@ -8,6 +8,7 @@ import * as twv from 'tailwind-variants'
 import * as aria from '#/components/aria'
 import * as ariaComponents from '#/components/AriaComponents'
 import * as portal from '#/components/Portal'
+import WindowButton from '#/components/styled/WindowButton'
```
Comment (MrFlashAccount) [inline app/ide-desktop/lib/dashboard/src/components/AriaComponents/Dialog/Dialog.tsx]: We have CloseButton, You don't have to use `WindowButton` or something

File: app/ide-desktop/lib/dashboard/src/components/AriaComponents/Dialog/Dialog.tsx [hunk 2]
```diff
@@ -150,6 +152,11 @@ export function Dialog(props: types.DialogProps) {
                   <div className="relative flex-auto overflow-y-auto p-3.5">
                     {typeof children === 'function' ? children(opts) : children}
                   </div>
+                  {closeButton === 'floating' && (
+                    <div className="m-floating-buttons absolute flex gap-1">
```
Comment (MrFlashAccount) [inline app/ide-desktop/lib/dashboard/src/components/AriaComponents/Dialog/Dialog.tsx]: Don't use `m-floating-buttons` or any other custom classes, because they make no sense and breaks style deduplication

File: app/ide-desktop/lib/dashboard/src/components/dashboard/ProjectIcon.tsx [hunk 3]
```diff
@@ -204,14 +184,35 @@ export default function ProjectIcon(props: ProjectIconProps) {
   )
 
   React.useEffect(() => {
```
Comment (MrFlashAccount) [inline app/ide-desktop/lib/dashboard/src/components/dashboard/ProjectIcon.tsx]: Please don't do async logic in useEffect. Use `useQuery` or `useMutation` instead.

File: app/ide-desktop/lib/dashboard/src/components/dashboard/ProjectIcon.tsx [hunk 4]
```diff
@@ -272,6 +266,15 @@ export default function ProjectIcon(props: ProjectIconProps) {
           setShouldSwitchPage(event.shouldAutomaticallySwitchPage)
           setIsRunningInBackground(event.runInBackground)
           void openProject(event.runInBackground)
+          void backend.getProjectDetails(item.id, item.parentId, item.title).then(project => {
```
Comment (MrFlashAccount) [inline app/ide-desktop/lib/dashboard/src/components/dashboard/ProjectIcon.tsx]: Do not fetch directly in event handles, use `react-query`

File: app/ide-desktop/lib/dashboard/src/components/dashboard/ProjectNameColumn.tsx [hunk 5]
```diff
@@ -73,10 +71,9 @@ export default function ProjectNameColumn(props: ProjectNameColumnProps) {
     (backend.type === backendModule.BackendType.local ||
       (ownPermission != null &&
         permissions.PERMISSION_ACTION_CAN_EXECUTE[ownPermission.permission]))
+  const isCloud = backend.type !== backendModule.BackendType.local
```
Comment (MrFlashAccount) [inline app/ide-desktop/lib/dashboard/src/components/dashboard/ProjectNameColumn.tsx]: This check doesn't mean that we're using cloud

File: app/ide-desktop/lib/dashboard/src/components/styled/Button.tsx [hunk 6]
```diff
@@ -62,13 +65,13 @@ function Button(props: ButtonProps, ref: React.ForwardedRef<HTMLButtonElement>)
...
           className: tailwindMerge.twMerge(
-            'relative after:pointer-events-none after:absolute after:inset-button-focus-ring-inset after:rounded-button-focus-ring',
+            'relative after:pointer-events-none after:absolute after:inset after:rounded-button-focus-ring transition-colors hover:enabled:bg-primary/10 rounded-button-focus-ring m-button-focus-ring-inset p-negative-button-focus-ring-inset',
             buttonClassName
           ),
         })}
       >
         <div
-          className={`group flex selectable ${isDisabled || softDisabled ? 'disabled' : ''} ${active ? 'active' : ''}`}
+          className={`group flex selectable ${light ? 'selectable-light' : ''} ${isDisabled || softDisabled ? 'disabled' : ''} ${active ? 'active' : ''}`}
```
Comment (MrFlashAccount) [inline app/ide-desktop/lib/dashboard/src/components/styled/Button.tsx]: Do not build class list using template literals, use either `tailwind-merge` or `clsx`

File: app/ide-desktop/lib/dashboard/src/components/styled/WindowButton.tsx [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+// ====================
+
+/** Props for a {@link WindowButton}. */
+export interface WindowButtonProps {
+  readonly onPress: () => void
+  readonly role: WindowButtonRole
+}
+
+/** A button for performing an action on a window. */
+export default function WindowButton(props: WindowButtonProps) {
```
Comment (MrFlashAccount) [inline app/ide-desktop/lib/dashboard/src/components/styled/WindowButton.tsx]: We already have `CloseButton` that does the same and looks better

File: app/ide-desktop/lib/dashboard/src/layouts/AssetsTable.tsx [hunk 8]
```diff
@@ -2469,7 +2489,26 @@ export default function AssetsTable(props: AssetsTableProps) {
...
+          onSelect={event => {
+            dispatchAssetListEvent({
+              type: AssetListEventType.uploadFiles,
+              parentKey: rootDirectoryId,
+              parentId: rootDirectoryId,
+              files: Array.from(event ?? []),
+            })
+          }}
+        >
+          <UnstyledButton
```
Comment (MrFlashAccount) [inline app/ide-desktop/lib/dashboard/src/layouts/AssetsTable.tsx]: Don't use `UnstyledButton`, use `ariaComponents.Button` with `custom` variant/size instead

File: app/ide-desktop/lib/dashboard/src/layouts/Drive.tsx [hunk 9]
```diff
@@ -221,50 +212,60 @@ export default function Drive(props: DriveProps) {
...
-      const newLabelName = backendModule.LabelName(value)
-      const placeholderLabel: backendModule.Label = {
-        id: backendModule.TagId(uniqueString.uniqueString()),
-        value: newLabelName,
-        color,
-      }
-      setNewLabelNames(labelNames => new Set([...labelNames, newLabelName]))
-      setLabels(oldLabels => [...oldLabels, placeholderLabel])
-      try {
-        const newLabel = await backend.createTag({ value, color })
```
Comment (MrFlashAccount) [inline app/ide-desktop/lib/dashboard/src/layouts/Drive.tsx:232]: Don't do mutations in `useCallback`. Use `useMutation` instead

File: app/ide-desktop/lib/dashboard/src/layouts/Drive.tsx [hunk 10]
```diff
@@ -221,50 +212,60 @@ export default function Drive(props: DriveProps) {
...
+        }
+        setNewLabelNames(labelNames => new Set([...labelNames, newLabelName]))
+        setLabels(oldLabels => [...oldLabels, placeholderLabel])
+        try {
+          const newLabel = await remoteBackend.createTag({ value, color })
+          setLabels(oldLabels =>
+            oldLabels.map(oldLabel => (oldLabel.id === placeholderLabel.id ? newLabel : oldLabel))
+          )
+        } catch (error) {
+          toastAndLog(null, error)
```
Comment (MrFlashAccount) [inline app/ide-desktop/lib/dashboard/src/layouts/Drive.tsx]: Do not toast errors.

File: app/ide-desktop/lib/dashboard/src/layouts/Drive.tsx [hunk 11]
```diff
@@ -221,50 +212,60 @@ export default function Drive(props: DriveProps) {
...
-      } catch (error) {
-        toastAndLog(null, error)
+      if (remoteBackend == null) {
+        // eslint-disable-next-line no-restricted-syntax
+        throw new Error('Labels can only be deleted on the Remote Backend.')
+      } else {
+        setDeletedLabelNames(oldNames => new Set([...oldNames, value]))
+        setQuery(oldQuery => oldQuery.deleteFromEveryTerm({ labels: [value] }))
+        try {
+          await remoteBackend.deleteTag(id, value)
```
Comment (MrFlashAccount) [inline app/ide-desktop/lib/dashboard/src/layouts/Drive.tsx]: do not fetch outside of `useQuery` / `useMutation`

File: app/ide-desktop/lib/dashboard/src/layouts/DriveBar.tsx [hunk 12]
```diff
@@ -112,28 +108,38 @@ export default function DriveBar(props: DriveBarProps) {
...
       )
     }
-    case Category.home: {
+    case Category.cloud:
+    case Category.local: {
       return (
         <div className="flex h-row py-drive-bar-y">
           <HorizontalMenuBar>
+            <aria.DialogTrigger>
+              <UnstyledButton
```
Comment (MrFlashAccount) [inline app/ide-desktop/lib/dashboard/src/layouts/DriveBar.tsx]: Don't use `UnstyledButton`

File: app/ide-desktop/lib/dashboard/src/layouts/Settings.tsx [hunk 13]
```diff
@@ -72,32 +72,37 @@ export default function Settings() {
     })()
   }, [sessionType, backend])
 
-  let content: React.JSX.Element
+  let content: JSX.Element | null
```
Comment (MrFlashAccount) [inline app/ide-desktop/lib/dashboard/src/layouts/Settings.tsx]: don't use `JSX.Element` since it's deprecated

File: app/ide-desktop/lib/dashboard/src/layouts/Settings.tsx [hunk 14]
```diff
@@ -72,32 +72,37 @@ export default function Settings() {
...
   switch (settingsTab) {
     case SettingsTab.account: {
-      content = <AccountSettingsTab />
+      content = backend == null ? null : <AccountSettingsTab backend={backend} />
```
Comment (MrFlashAccount) [inline app/ide-desktop/lib/dashboard/src/layouts/Settings.tsx:57]: If backend is null then show loader. Don't pass null.

File: app/ide-desktop/lib/dashboard/src/layouts/Settings/DeleteUserAccountSettingsSection.tsx [hunk 15]
```diff
@@ -12,15 +11,24 @@ import UnstyledButton from '#/components/UnstyledButton'
...
 
+import type Backend from '#/services/Backend'
+
 // ========================================
 // === DeleteUserAccountSettingsSection ===
 // ========================================
 
+/** Props for a {@link DeleteUserAccountSettingsSection}. */
+export interface DeleteUserAccountSettingsSectionProps {
+  readonly backend: Backend
```
Comment (MrFlashAccount) [inline app/ide-desktop/lib/dashboard/src/layouts/Settings/DeleteUserAccountSettingsSection.tsx:24]: Why do we pass backend instead of consuming it from the `Context`? Props drilling is a bad practice.

File: app/ide-desktop/lib/dashboard/src/layouts/Settings/MembersSettingsTab.tsx [hunk 16]
```diff
@@ -15,94 +15,46 @@ import SettingsSection from '#/components/styled/settings/SettingsSection'
...
-        staleTime: LIST_USERS_STALE_TIME,
+        staleTime: LIST_USERS_STALE_TIME_MS,
       },
     ],
   })
 
   return (
     <SettingsPage>
       <SettingsSection noFocusArea title={getText('members')} className="overflow-hidden">
-        <MembersSettingsTabBar />
```
Comment (MrFlashAccount) [inline app/ide-desktop/lib/dashboard/src/layouts/Settings/MembersSettingsTab.tsx]: Did you just remove the `invitations` ? Please revert the deletion

File: app/ide-desktop/lib/dashboard/src/layouts/Settings/MembersTable.tsx [hunk 17]
```diff
@@ -32,25 +33,23 @@ export interface MembersTableProps {
...
   const scrollContainerRef = React.useRef<HTMLDivElement>(null)
   const bodyRef = React.useRef<HTMLTableSectionElement>(null)
-  const members = asyncEffectHooks.useAsyncEffect<backendModule.User[] | null>(
-    !populateWithSelf || user == null ? null : [user],
-    () => backend.listUsers(),
-    [backend]
-  )
+  const membersQuery = reactQuery.useQuery({
+    initialData: () => (populateWithSelf && user != null ? [user] : null),
+    queryKey: ['members'],
```
Comment (MrFlashAccount) [inline app/ide-desktop/lib/dashboard/src/layouts/Settings/MembersTable.tsx]: use `listUsers` key instead

File: app/ide-desktop/lib/dashboard/src/layouts/Start.tsx [hunk 18]
```diff
@@ -0,0 +1,49 @@
...
+// === Home ===
+// ============
+
+/** Props for a {@link Home}. */
+export interface HomeProps {
+  readonly createProject: (templateId?: string | null, templateName?: string | null) => void
+}
+
+/** Home screen. */
+export default function Home(props: HomeProps) {
```
Comment (MrFlashAccount) [inline app/ide-desktop/lib/dashboard/src/layouts/Start.tsx]: If it's a dialog, why it's in layouts? Also the name doesn't represent tha this is a dialog component

File: app/ide-desktop/lib/dashboard/src/providers/BackendProvider.tsx [hunk 19]
```diff
@@ -52,32 +35,74 @@ export interface BackendProviderProps extends Readonly<React.PropsWithChildren>
...
+// ==============================
+
+/** Get the Remote Backend.
+ * @throws {Error} when no Remote Backend exists. This should only happen if the user is not logged
+ * in. */
+export function useRemoteBackendStrict() {
+  const remoteBackend = React.useContext(BackendContext).remoteBackend
+  if (remoteBackend == null) {
+    // eslint-disable-next-line no-restricted-syntax
+    throw new Error('This component requires a Cloud Backend to function.')
```
Comment (MrFlashAccount) [inline app/ide-desktop/lib/dashboard/src/providers/BackendProvider.tsx:70]: use `invariant` instead

File: app/ide-desktop/lib/dashboard/src/providers/BackendProvider.tsx [hunk 20]
```diff
@@ -52,32 +35,74 @@ export interface BackendProviderProps extends Readonly<React.PropsWithChildren>
...
+/** Get the corresponding backend for the given property.
+ * @throws {Error} when neither the Remote Backend nor the Local Backend are supported.
+ * This should never happen unless the build is misconfigured. */
+export function useBackend(category: Category) {
+  const remoteBackend = useRemoteBackend()
+  const localBackend = useLocalBackend()
+  if (categoryModule.isCloud(category)) {
+    if (remoteBackend == null) {
+      // eslint-disable-next-line no-restricted-syntax
+      throw new Error(
```
Comment (MrFlashAccount) [inline app/ide-desktop/lib/dashboard/src/providers/BackendProvider.tsx]: use invariant instead

File: app/ide-desktop/lib/dashboard/src/services/RemoteBackend.ts [hunk 21]
```diff
@@ -71,6 +71,7 @@ export async function waitUntilProjectIsReady(
     await remoteBackend.openProject(item.id, null, item.title)
   }
   let nextCheckTimestamp = 0
+  console.log(abortController.signal.aborted, project.state.type)
```
Comment (MrFlashAccount) [inline app/ide-desktop/lib/dashboard/src/services/RemoteBackend.ts]: Remove please console.log

File: app/ide-desktop/lib/dashboard/src/tailwind.css [hunk 22]
```diff
@@ -156,11 +156,12 @@
   --auth-link-padding-x: 0.5rem;
   --auth-link-padding-y: 0.25rem;
   --text-link-padding-x: 0.25rem;
+  /* The margin between a floating button and the edge of its window. */
+  --floating-buttons-margin: 1.1875rem;
```
Comment (MrFlashAccount) [inline app/ide-desktop/lib/dashboard/src/tailwind.css]: Don't use custom margins

File: app/ide-desktop/lib/dashboard/tailwind.config.js [hunk 23]
```diff
@@ -19,11 +19,14 @@ export default /** @satisfies {import('tailwindcss').Config} */ ({
...
         // This should be named "regular".
         primary: 'rgb(0 0 0 / 60%)',
+        accent: '#499f4b',
         'hover-bg': 'rgb(0 0 0 / 10%)',
         frame: 'var(--frame-color)',
         'selected-frame': 'var(--selected-frame-color)',
         'ide-bg': '#ebeef1',
         selected: 'rgb(255 255 255 / 40%)',
+        'not-selected': 'rgb(0 0 0 / 15%)',
+        'window-close': '#ff605c',
```
Comment (MrFlashAccount) [inline app/ide-desktop/lib/dashboard/tailwind.config.js]: use generic names instead of specific names

File: app/ide-desktop/lib/dashboard/tailwind.config.js [hunk 24]
```diff
@@ -94,6 +98,7 @@ export default /** @satisfies {import('tailwindcss').Config} */ ({
         'chat-profile-picture': 'var(--chat-profile-picture-size)',
         'selection-brush-border': 'var(--selection-brush-border-width)',
         'button-focus-ring-inset': 'var(--button-focus-ring-inset)',
+        'negative-button-focus-ring-inset': 'calc(var(--button-focus-ring-inset) * -1)',
```
Comment (MrFlashAccount) [inline app/ide-desktop/lib/dashboard/tailwind.config.js]: don't use custom spacings

File: app/ide-desktop/lib/dashboard/tailwind.config.js [hunk 25]
```diff
@@ -345,6 +346,7 @@ export default /** @satisfies {import('tailwindcss').Config} */ ({
         'date-input-calendar-gap': 'var(--date-input-calendar-gap)',
         'context-menu-macos-half-x': 'var(--context-menu-macos-half-width)',
         'context-menu-half-x': 'var(--context-menu-half-width)',
+        'floating-buttons': 'var(--floating-buttons-margin)',
```
Comment (MrFlashAccount) [inline app/ide-desktop/lib/dashboard/tailwind.config.js]: use only generic classes

File: app/ide-desktop/lib/dashboard/tailwind.config.js [hunk 26]
```diff
@@ -498,6 +501,9 @@ inset 0 -36px 51px -51px #00000014`,
             '@apply disabled:opacity-30 [&.disabled]:opacity-30 disabled:cursor-not-allowed [&.disabled]:cursor-not-allowed opacity-50 hover:opacity-75 transition-all':
               '',
           },
+          '.selectable-light': {
+            '@apply opacity-25': '',
```
Comment (MrFlashAccount) [inline app/ide-desktop/lib/dashboard/tailwind.config.js]: don't use appy and custom classes unless you have to

**Review by MrFlashAccount — CHANGES_REQUESTED** [review]

File: app/ide-desktop/lib/dashboard/src/components/dashboard/ProjectNameColumn.tsx (hunk 5 above)
Comment (somebody1234) [reply app/ide-desktop/lib/dashboard/src/components/dashboard/ProjectNameColumn.tsx]: oh?

File: app/ide-desktop/lib/dashboard/src/layouts/Settings.tsx (hunk 14 above)
Comment (somebody1234) [reply app/ide-desktop/lib/dashboard/src/layouts/Settings.tsx:57]: loader makes no sense here, if backend is null we cannot show the tab at all

File: app/ide-desktop/lib/dashboard/src/layouts/Settings/DeleteUserAccountSettingsSection.tsx (hunk 15 above)
Comment (somebody1234) [reply app/ide-desktop/lib/dashboard/src/layouts/Settings/DeleteUserAccountSettingsSection.tsx:24]: this is not props drilling, it is one level deep. putting everything in a context to hide state is a bad practice

File: app/ide-desktop/lib/dashboard/tailwind.config.js (hunk 23 above)
Comment (somebody1234) [reply app/ide-desktop/lib/dashboard/tailwind.config.js]: this is the generic name, this is the macos close color

File: app/ide-desktop/lib/dashboard/src/components/dashboard/ProjectNameColumn.tsx (hunk 5 above)
Comment (MrFlashAccount) [reply app/ide-desktop/lib/dashboard/src/components/dashboard/ProjectNameColumn.tsx]: `backend.type !== backendModule.BackendType.local` means that we're using a backend that is not local.
Use `backend.type === backendModule.BackendType.remote` instead

**Comment by MrFlashAccount** [comment]
QA ⛔️

1. Can we improve the drag experience to look similar to Google Drive? Currently, if user have a lot of assets, the DND button is hidden.
https://github.com/enso-org/enso/assets/61194245/066d2aa7-cbfe-415f-88ac-2e8b4738fa43
2. Invitations disappeared in settings
3. Buttons are not aligned (use `ariaComponents.Button`)
CleanShot 2024-05-27 at 12 49 30

File: app/ide-desktop/lib/dashboard/src/components/AriaComponents/Dialog/Dialog.tsx (hunk 2 above)
Comment (somebody1234) [reply app/ide-desktop/lib/dashboard/src/components/AriaComponents/Dialog/Dialog.tsx]: style deduplication should not be a concern though

Comment (MrFlashAccount) [reply app/ide-desktop/lib/dashboard/src/components/AriaComponents/Dialog/Dialog.tsx]: It's always a concern

File: app/ide-desktop/lib/dashboard/src/layouts/Settings.tsx (hunk 14 above)
Comment (MrFlashAccount) [reply app/ide-desktop/lib/dashboard/src/layouts/Settings.tsx:57]: When the backend could be null?

Comment (MrFlashAccount) [reply app/ide-desktop/lib/dashboard/src/layouts/Settings.tsx:57]: > loader makes no sense here, if backend is null we cannot show the tab at all

Do we exclude the tab from the sidebar then?

File: app/ide-desktop/lib/dashboard/src/layouts/Settings/DeleteUserAccountSettingsSection.tsx (hunk 15 above)
Comment (MrFlashAccount) [reply app/ide-desktop/lib/dashboard/src/layouts/Settings/DeleteUserAccountSettingsSection.tsx:24]: If we need backend deeper inside the React components tree, do we need to pass it as a prop or we can use context instead?

File: app/ide-desktop/lib/dashboard/src/components/AriaComponents/Dialog/Dialog.tsx (hunk 2 above)
Comment (somebody1234) [reply app/ide-desktop/lib/dashboard/src/components/AriaComponents/Dialog/Dialog.tsx]: but why?

File: app/ide-desktop/lib/dashboard/src/layouts/Settings.tsx (hunk 14 above)
Comment (somebody1234) [reply app/ide-desktop/lib/dashboard/src/layouts/Settings.tsx:57]: @MrFlashAccount currently yes (iirc). in the future we will probably want to show it under a paywall instead.

File: app/ide-desktop/lib/dashboard/src/layouts/Settings/DeleteUserAccountSettingsSection.tsx (hunk 15 above)
Comment (somebody1234) [reply app/ide-desktop/lib/dashboard/src/layouts/Settings/DeleteUserAccountSettingsSection.tsx:24]: typically context (like how it is used everywhere else). i just figured that here specifically the root `Settings` component should be controlling basically all the state of its children

Comment (somebody1234) [reply app/ide-desktop/lib/dashboard/src/layouts/Settings/DeleteUserAccountSettingsSection.tsx:24]: the settings search PR is probably vaguely related

**Comment by somebody1234** [comment]
re: 3 - most style issues were fixed in the second PR. i figured it's easier to implement it once rather than having to backport. thoughts?

File: app/ide-desktop/lib/dashboard/src/layouts/Settings.tsx (hunk 14 above)
Comment (MrFlashAccount) [reply app/ide-desktop/lib/dashboard/src/layouts/Settings.tsx:57]: How's it related?

File: app/ide-desktop/lib/dashboard/src/components/AriaComponents/Dialog/Dialog.tsx (hunk 2 above)
Comment (MrFlashAccount) [reply app/ide-desktop/lib/dashboard/src/components/AriaComponents/Dialog/Dialog.tsx]: Because:

1. `!important` hell
2. m-floating-buttons - how'd I understand how much it is in pixels?
3. tw-merge doesn't know `m-floating-buttons` 
4. We use the atomic approach, and style deduplication is one of the most important features.
5. We have tailwind-variants - if you want to add styles like 'button' use this lib instead.

File: app/ide-desktop/lib/dashboard/src/layouts/Settings/DeleteUserAccountSettingsSection.tsx (hunk 15 above)
Comment (MrFlashAccount) [reply app/ide-desktop/lib/dashboard/src/layouts/Settings/DeleteUserAccountSettingsSection.tsx:24]: So why do we pass the backend directly then?

Comment (MrFlashAccount) [reply app/ide-desktop/lib/dashboard/src/layouts/Settings/DeleteUserAccountSettingsSection.tsx:24]: Looks like a stylistic choice that makes no sense and leads to unneeded cognitive load and merge conflicts.
Like a change only to make a change.

File: app/ide-desktop/lib/dashboard/src/components/AriaComponents/Dialog/Dialog.tsx (hunk 2 above)
Comment (somebody1234) [reply app/ide-desktop/lib/dashboard/src/components/AriaComponents/Dialog/Dialog.tsx]: > `!important` hell

should not happen, an `m-` is still an `m-`.

> how'd I understand how much it is in pixels?

you're not supposed to

> tw-merge doesn't know `m-floating-buttons`

however `tw-merge` is trivial to configure so this is not an issue in practice

> style deduplication is one of the most important features

not too sure about this. because of variants, we probably end up with much more generated CSS.

> We have `tailwind-variants`

not sure how this is different to (or better than) using `@apply` directly


also remember: the goal of tailwind is not to golf your css. the goal of tailwind is not to deduplicate styles. the goal of tailwind is to encode a design system, and arbitrary pixel values are not a design system.

Comment (MrFlashAccount) [reply app/ide-desktop/lib/dashboard/src/components/AriaComponents/Dialog/Dialog.tsx]: > not too sure about this. because of variants, we probably end up with much more generated CSS.

It's not

> not sure how this is different to (or better than) using @apply directly

Apply is considered a bad practice by the authors of tw.

> you're not supposed to

Yes we supposed to. Because we. read. the. code.

Comment (MrFlashAccount) [reply app/ide-desktop/lib/dashboard/src/components/AriaComponents/Dialog/Dialog.tsx]: We're not going to use `m-floating-buttons`, `button` or any other classes. Period

File: app/ide-desktop/lib/dashboard/src/layouts/Settings.tsx (hunk 14 above)
Comment (somebody1234) [reply app/ide-desktop/lib/dashboard/src/layouts/Settings.tsx:78]: in the PR the backend is refactored out to [a context object](https://github.com/enso-org/enso/blob/377e23da4a29cebc907ac2abb244b73e82133615/app/ide-desktop/lib/dashboard/src/layouts/Settings/settingsData.tsx#L374)

Comment (somebody1234) [reply app/ide-desktop/lib/dashboard/src/layouts/Settings.tsx:57]: how's what related?

File: app/ide-desktop/lib/dashboard/src/layouts/Settings/DeleteUserAccountSettingsSection.tsx (hunk 15 above)
Comment (somebody1234) [reply app/ide-desktop/lib/dashboard/src/layouts/Settings/DeleteUserAccountSettingsSection.tsx:24]: > unneeded cognitive load 

imo contexts are a higher cognitive load since it is much harder to track where the state is coming from, let alone what its lifecycle looks like (in contrast, with props it's obvious where the state comes from - the parent). if you have a different opinion please do elaborate

> merge conflicts

explain

Comment (somebody1234) [reply app/ide-desktop/lib/dashboard/src/layouts/Settings/DeleteUserAccountSettingsSection.tsx:24]: re: merge conflicts specifically - this is something that should never change, so i'm not sure whether this will be a problem (like at all)

also note that the main *reason* why this change is happening is so that the main `Settings` component controls all the state that is passed to each of its children. this means that the inputs are only ever acquired in one place, so that if the inputs change, only the hook in the main `Settings.tsx` needs to be changed, not every such occurrence of the hook in the entire `Settings/` subfolder using a hacky regex search+replace.

**Comment by somebody1234** [comment]
2 & 3. should be addressed.
1. seems like it's a bigger feature, but *hopefully* it's the last thing to address (after some other final refactors)

**Comment by somebody1234** [comment]
on second thought 1 will be left for later

File: app/ide-desktop/lib/dashboard/src/layouts/Settings.tsx (hunk 14 above)
Comment (MrFlashAccount) [reply app/ide-desktop/lib/dashboard/src/layouts/Settings.tsx:57]: I have no idea how backend == null is related to paywalls

Comment (MrFlashAccount) [reply app/ide-desktop/lib/dashboard/src/layouts/Settings.tsx:57]: And at this point I don't understand why the backend could be null in settings

Comment (somebody1234) [reply app/ide-desktop/lib/dashboard/src/layouts/Settings.tsx:57]: @MrFlashAccount true... i was thinking that the tabs *could* be visible for the purpose of showing paywalls. but of course, if `backend == null` then the user is not logged in (so we don't know the user's plan, so no paywalls)

Comment (MrFlashAccount) [reply app/ide-desktop/lib/dashboard/src/layouts/Settings.tsx:57]: Is it possible now to show settings without logging in? Also, I think this case should be handled by the router(if a user is not logged in, redirect him to the login screen). WDYT?

Comment (somebody1234) [reply app/ide-desktop/lib/dashboard/src/layouts/Settings.tsx:57]: > Is it possible now to show settings without logging in?

no idea. i guess *technically* no because it's no longer possible to avoid logging in?

> Also, I think this case should be handled by the router(if a user is not logged in, redirect him to the login screen). WDYT?

depends on whether we ever want to enable the app for users that are not logged in (e.g. custom builds without the cloud backend). not sure what happens with an *expired* session token when the user is physically offline, but we may need to take that into account too.

so tl;dr: *assuming we will never ever need to show settings without logging in* then it makes sense to `invariant` this. otherwise i don't think it hurts to retain this logic.

it's possible that we may be able to make this tidier by using a `Router` and `Layout`s here? thoughts?

Comment (MrFlashAccount) [reply app/ide-desktop/lib/dashboard/src/layouts/Settings.tsx:57]: > it's possible that we may be able to make this tidier by using a Router and Layouts here? thoughts?

CleanShot 2024-05-29 at 12 40 24@2x 
CleanShot 2024-05-29 at 12 41 06@2x

Looks like this is exactly how it's working right now.

Comment (MrFlashAccount) [reply app/ide-desktop/lib/dashboard/src/layouts/Settings.tsx:57]: > no idea. i guess technically no because it's no longer possible to avoid logging in?

Yeah, seems like it's not possible anymore

Comment (MrFlashAccount) [reply app/ide-desktop/lib/dashboard/src/layouts/Settings.tsx:57]: > depends on whether we ever want to enable the app for users that are not logged in (e.g. custom builds without the cloud backend). not sure what happens with an expired session token when the user is physically offline, but we may need to take that into account too.


Yeah, the behavior of the dashboard, when a user goes offline is still unclear... I think we need to revisit this part one day.

Comment (somebody1234) [reply app/ide-desktop/lib/dashboard/src/layouts/Settings.tsx:57]: (re: routes)

right, but `remoteBackend` may still be `null` because the user might be in offline mode (`session.type === 'offline'`). not sure where it stands today (see above re: using the app offline inside e.g. company proxy)

so the `Layout`s here would be to distinguish between settings tab that need the backend and those that don't

**Comment by PabloBuchu** [comment]
1. After clicking `Start with a template` and closing modal I end up with this
Screenshot 2024-06-10 at 13 40 47

2. `New empty project` and polling for status is bugged
Screenshot 2024-06-10 at 13 41 18
Screenshot 2024-06-10 at 13 41 12

3. Should this be so weird
Screenshot 2024-06-10 at 13 43 06

4. Tooltips problem
https://github.com/enso-org/enso/assets/2855109/db6802ae-ded6-4b8d-b7ec-f5a96985ae9c

5. Show secret value is not aligned and doesn't work
Screenshot 2024-06-10 at 13 44 56

6. Share with modal
Screenshot 2024-06-10 at 13 46 05

7. Have we disabled cmd / shift + arrows multiple selection?

**Comment by somebody1234** [comment]
uhoh

**Comment by somebody1234** [comment]
fixed some things locally, some of them are still TODO:
1. fixed, assuming the issue is the focus ring being cut off
  - i've noticed the focus ring is inconsistent (spacing, animations) - this should probably be fixed at some point as well
2. TODO
3. fixed
4. fixed, assuming the issue is the positioning causing issues
5. fixed
6. fixed
7. (debugging TODO) idts - but it seems to be bugged, i can see the selection flashing for a moment before being reset to just the one row

**Comment by PabloBuchu** [comment]
@somebody1234 we can fix 7th on separate PR and then just fix 2nd issue here. Then we can push this to merge after all others are fixed

**Comment by somebody1234** [comment]
found the cause of 2nd, was me being silly. @PabloBuchu should be ready for re-QA now - although now's probably not the best time, lol

**Comment by somebody1234** [comment]
@MrFlashAccount just a note that this PR is blocked on your re-review

**Comment by PabloBuchu** [comment]
QA passed. I created one small fix https://github.com/enso-org/cloud-v2/issues/1301

lets fix the merge conflict and try to merge it today

@MrFlashAccount can you rereview and approve?

**Review by PabloBuchu — APPROVED** [review]

**Review by MrFlashAccount — APPROVED** [review]
